### PR TITLE
Export omeka

### LINKED ISSUE #100: Update actions/upload-artifact to v4
### Describe what feature you'd like. Pseudo-code, mockups, or screenshots of similar solutions are encouraged!

Update actions/upload-artifact to v4 in ci pipeline.

See 
Screenshot 2024-05-15 at 09 45 15

### What type of pull request would this be?

Enhancement

### Any links to similar examples or other references we should review?

See https://github.com/Stadt-Geschichte-Basel/forschung.stadtgeschichtebasel.ch/actions/runs/8664294518

### LINKED ISSUE #106: Verify Rights Before Offering Original File Downloads
It's important to ensure that all rights are verified before offering original file downloads in the CSV output. Documentation regarding this decision will be added to clarify the process and requirements. This issue was requested to be opened by @maehr in PR https://github.com/Stadt-Geschichte-Basel/forschung.stadtgeschichtebasel.ch/pull/101.

### PR DESCRIPTION
**Pull request by maehr**
# Pull request

## Proposed changes

- Update metadata model
- Fix #100 
- Fix #106 

## Types of changes

- [ ] New feature (non-breaking change which adds functionality).
- [x] Enhancement (non-breaking change which enhances functionality)
- [x] Bug Fix (non-breaking change which fixes an issue).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

## Checklist

- [ ] I have read the **[CONTRIBUTING](./CONTRIBUTING.md)** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by maehr** [comment]
@coderabbitai Read #100 again. It should have said update to v4.

File: .github/workflows/process_data.py [hunk 1]
```diff
@@ -51,105 +51,174 @@ def get_media(item_id):
...
-        "license": data.get("dcterms:license", [{}])[0].get("@value"),
-        "isPartOf": data.get("dcterms:isPartOf", [{}])[0].get("@value"),
-        "isReferencedBy": data.get("dcterms:isReferencedBy", [{}])[0].get("@value"),
-        "display_template": type,
-        "object_location": media_original_url,
-        "image_small": media_url,
-        "image_thumb": media_url,
-        "image_alt_text": None,
-        "object_transcript": None,
+        "objectid": item["o:id"],
```
Comment (koilebeit) [inline .github/workflows/process_data.py:115]: you are using the omeka internal id here. but it needs "our" id (dcterms:identifier)
```suggestion
        "objectid": extract_prop_value(item.get("dcterms:identifier", []), 10),
```
but of course this has consequences for line 201

File: .github/workflows/process_data.py [hunk 2]
```diff
@@ -51,105 +51,174 @@ def get_media(item_id):
...
+    display_template = infer_display_template(mime_type)
+
+    image_url = media.get("thumbnail_display_urls", {}).get("large", "")
+    local_image_path = ""
+    if image_url:
+        local_image_path = f"objects/{media['o:id']}.jpg"
+        download_file(image_url, local_image_path)
+
+    return {
+        "objectid": media["o:id"],
```
Comment (koilebeit) [inline .github/workflows/process_data.py:163]: see comment above
```suggestion
        "objectid": extract_prop_value(media.get("dcterms:identifier", []), 10),
```

File: .github/workflows/process_data.py [hunk 3]
```diff
@@ -51,105 +51,174 @@ def get_media(item_id):
...
+    collection_id = ITEM_SET_ID
+    items_data = get_items_from_collection(collection_id)
+
+    # Extract item data
+    item_records = [extract_item_data(item) for item in items_data]
+
+    # Extract media data for each item
+    media_records = []
+    for item in items_data:
+        item_id = item["o:id"]
```
Comment (koilebeit) [inline .github/workflows/process_data.py:201]: see comment above

**Review by koilebeit — CHANGES_REQUESTED** [review]
You use the omeka internal id. But as id we need the abb or m identifier, don't we? I don't know if it's just that, but the item pages are not found this way...

**Comment by maehr** [comment]
@koilebeit I'll try to fix this asap. Shouldn't be a biggy

**Comment by maehr** [comment]
@koilebeit the csv must not contain quotation marks. Somehow the gallery worked but browse didn't. Thats why I didn't notice the bug. Now it builds locally.

File: .github/workflows/process_data.py [hunk 4]
```diff
@@ -51,105 +51,187 @@ def get_media(item_id):
...
+        "date": extract_prop_value(media.get("dcterms:date", []), 7),
+        "type": extract_prop_uri(media.get("dcterms:type", []), 8),
+        "format": format_value,
+        "extent": extract_prop_value(media.get("dcterms:extent", []), 25),
+        "language": extract_prop_value(media.get("dcterms:language", []), 12),
+        "relation": extract_combined_list(media.get("dcterms:relation", [])),
+        "rights": extract_prop_value(media.get("dcterms:rights", []), 15),
+        "license": extract_prop_value(media.get("dcterms:license", []), 49),
+        "display_template": display_template,
+        "object_location": local_image_path,
```
[the remark this answers is not in the document]
Comment (maehr) [reply .github/workflows/process_data.py:195]: I did this on purpose. We should make sure first to have all the rights before we offer the originals. I'll add some documentation.
@coderabbitai open an issue for this

**Review by koilebeit — APPROVED** [review]
When we merge, however, only the objects from the new collection will be online

**Comment by koilebeit** [comment]
If it is possible, the object_location should at least be specified correctly for the geodata, otherwise the new map view (in #99) will not work

**Comment by maehr** [comment]
> When we merge, however, only the objects from the new collection will be online. And the display (especially the links) still contains the markdown

Do you want to fix this in the view rather than here?

**Comment by koilebeit** [comment]
I noticed two more things:
1. the parent objects do not need an object_location. Should be a quick fix.
2. how do we deal with media appearing in several parent objects (e.g. the placeholder image)? In omeka this is easy, but in CB-CSV each objectid must be unique and can only have one parentid (see [https://collectionbuilder.github.io/cb-docs/docs/metadata/compound-objects/#metadata-conventions](https://collectionbuilder.github.io/cb-docs/docs/metadata/compound-objects/#metadata-conventions)). So in cb no n-m relationship is possible, only a 1-n relationship.

**Comment by maehr** [comment]
> I noticed two more things:
> 
> 1. the parent objects do not need an object_location. Should be a quick fix.
> 2. how do we deal with media appearing in several parent objects (e.g. the placeholder image)? In omeka this is easy (although I have no idea how to do this in the omeka interface), but in CB-CSV each objectid must be unique and can only have one parentid (see https://collectionbuilder.github.io/cb-docs/docs/metadata/compound-objects/#metadata-conventions). So in cb no n-m relationship is possible, only a 1-n relationship.

Thanks for pointing out these bugs. I'll try something

**Comment by maehr** [comment]
@koilebeit I fixed bug 1 and I figured out that even though the docs say differently, you can just add a second line copying over the media object and adjust the parent_id. It works. (It's not a nice workaround, but it serves the purpose, right?)

**Pull request merged**